### PR TITLE
Add Undo/Redo commands

### LINKED ISSUE #91: As a careless user, I can undo commands
... easily fix any errors I might make.

### LINKED ISSUE #92: As a careless user, I can redo commands
... easily fix an erroneous undo.

### PR DESCRIPTION
**Pull request by benedictkhoomw**
## What this does
This PR
1. adds undo/redo capabilities to SunRez
2. adds GUI menu items for undo/redo that enable/disable based on the state of SunRez
3. adds keyboard shortcuts for undo/redo (`CTRL + Z`/`CMD + Z` and `CTRL + SHIFT + Z`/`CMD + SHIFT + Z` respectively)
4. adds UG entries documenting these features

Closes #91 
Closes #92 

## How to test
1. Launch SunRez.
2. Check the `Edit` menu item. The sub-menu items `Undo` and `Redo` should both be disabled.
2. Perform some state-changing commands, but not `alias` (not supported yet).
3. Check the `Edit` menu item again. `Undo` should be enabled now.
3. Execute the `undo` command. The command should be undone.
4. Check the `Edit` menu item again. `Redo` should be enabled now.
4. Execute the `redo` command. The command just undone should be re-executed.
5. Press `CTRL + Z` (Windows, Linux) or `CMD + Z` (Mac) or try the relevant menu options at the top of the SunRez window. This should have the same effect as a regular `undo` command, including adding `undo` and `redo` to command history.
6. Repeat steps 2-8 any number of times and in any order. An error message should be shown if `undo` is executed when there are no more commands to undo, or if `redo` is executed when there are no more commands to redo. Otherwise, a success message should be shown.

## What is left to be done
- Integrate alias command
- Write tests
- Update DG

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/undoredo/RedoCommand.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                || other instanceof RedoCommand; // instanceof handles nulls
+    }
+
+    /**
+     * Undoes the most recent undo command, if any.
+     *
+     * @param model {@code Model} which the command should operate on.
+     * @return A {@code CommandResult} with a success message.
+     * @throws CommandException If the redo fails for whatever reason (e.g. no next state).
+     */
```
Comment (cnlinh) [inline src/main/java/seedu/address/logic/commands/undoredo/RedoCommand.java:38]: @throws NullPointerException?

File: src/main/java/seedu/address/logic/commands/undoredo/UndoCommand.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                || other instanceof UndoCommand; // instanceof handles nulls
+    }
+
+    /**
+     * Undoes the most recent state-changing command, if any.
+     *
+     * @param model {@code Model} which the command should operate on.
+     * @return A {@code CommandResult} with a success message.
+     * @throws CommandException If the undo fails for whatever reason (e.g. no previous state).
+     */
```
Comment (cnlinh) [inline src/main/java/seedu/address/logic/commands/undoredo/UndoCommand.java:38]: @throws NullPointerException?

**Review by colintkn — COMMENTED** [review]
> adds keyboard shortcuts for undo/redo (CTRL + Z/CMD + Z and CTRL + SHIFT + Z/CMD + SHIFT + Z respectively)

Currently on Mac, the undo keystroke is `CTRL + Z` and redo keystroke is `CTRL +SHIFT +Z` 
instead of `CMD + Z` and `CMD + SHIFT + Z`. 
 
Not a big problem. Feature works perfect other than that.

File: src/main/java/seedu/address/logic/commands/undoredo/RedoCommand.java (hunk 1 above)
Comment (benedictkhoomw) [reply src/main/java/seedu/address/logic/commands/undoredo/RedoCommand.java:39]: Thanks, added.

File: src/main/java/seedu/address/logic/commands/undoredo/UndoCommand.java (hunk 2 above)
Comment (benedictkhoomw) [reply src/main/java/seedu/address/logic/commands/undoredo/UndoCommand.java:39]: Thanks, added.

**Review by cnlinh — APPROVED** [review]
LGTM

**Review by DrWala — APPROVED** [review]
LGTM

**Pull request merged**